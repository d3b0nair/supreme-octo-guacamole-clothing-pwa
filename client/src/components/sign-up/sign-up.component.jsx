import React, { useState, useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectUserError } from "../../redux/user/user.selector";
import { signUpStart } from "../../redux/user/user.actions";

import FormInput from "../form-input/form-input.component";
import CustomButton from "../custom-button/custom-button.component";
import {
  SignUpContainer,
  TitleContainer,
  VisibleIconWrapper,
  HideIconWrapper,
  Wrapper,
  TextMsg,
  TextContainer,
} from "./sign-up.styles.jsx";

const ErrorMsgOnEmailAlreadyInUse = ({ emailAlreadyInUse }) =>
  emailAlreadyInUse ? (
    <TextContainer>
      <TextMsg>❌ Email is already in use</TextMsg>
    </TextContainer>
  ) : null;

const ErrorMsgOnPasswordDoesntMatch = ({ passwordDontMatch }) =>
  passwordDontMatch ? (
    <TextContainer>
      <TextMsg>❌ Passwords doesn't match</TextMsg>
    </TextContainer>
  ) : null;

const passwordRegex = "^(?=.*)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$";

const PasswordStrenghtField = ({ password }) => {
  const eightCharactersRegex = new RegExp("[a-zA-Z]+.{8,}").test(`${password}`);
  const capitalLetterRegex = new RegExp("[A-Z]").test(`${password}`);
  const lowerCaseLetterRegex = new RegExp("[a-z]").test(`${password}`);
  const specialCharRegex = new RegExp("[!@#$&*]").test(`${password}`);
  const numRegex = new RegExp("[0-9]").test(`${password}`);
  return password.length > 0 ? (
    <TextContainer>
      <TextMsg style={eightCharactersRegex ? { color: "#299329de" } : null}>
        {eightCharactersRegex ? "✔️" : "❌"} At least 8 characters long
      </TextMsg>
      <TextMsg style={capitalLetterRegex ? { color: "#299329de" } : null}>
        {capitalLetterRegex ? "✔️" : "❌"} At least 1 capital letter
      </TextMsg>
      <TextMsg style={lowerCaseLetterRegex ? { color: "#299329de" } : null}>
        {lowerCaseLetterRegex ? "✔️" : "❌"} At least 1 lowercase letter
      </TextMsg>
      <TextMsg style={specialCharRegex ? { color: "#299329de" } : null}>
        {specialCharRegex ? "✔️" : "❌"} At least 1 special character
      </TextMsg>
      <TextMsg style={numRegex ? { color: "#299329de" } : null}>
        {numRegex ? "✔️" : "❌"} At least 1 numeric character
      </TextMsg>
    </TextContainer>
  ) : null;
};

export function SignUp({ signUpStart, userError }) {
  const [newUser, setNewUser] = useState({
    displayName: "",
    email: "",
    password: "",
    confirmPassword: "",
  });
  const [passwordDontMatch, setPasswordDontMatch] = useState(null);
  const [showPasswordToggle, setShowPasswordToggle] = useState(false);
  const [emailAlreadyInUse, setEmailAlreadyInUse] = useState(null);
  useEffect(() => {
    if (newUser.password !== newUser.confirmPassword) {
      setPasswordDontMatch(true);
      return;
    } else if (newUser.password === newUser.confirmPassword) {
      setPasswordDontMatch(false);
    }
  }, [newUser.confirmPassword, newUser.password]);
  useEffect(() => {
    if (userError && userError.code === "auth/email-already-in-use") {
      setEmailAlreadyInUse(true);
    }
  }, [userError]);
  const handleSubmit = async (e) => {
    const { displayName, email, password } = newUser;
    e.preventDefault();

    await signUpStart({ displayName, email, password });
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setNewUser({ ...newUser, [name]: value });
  };

  const ShowAndHidePasswordToggle = () =>
    showPasswordToggle ? (
      <VisibleIconWrapper
        onClick={() => setShowPasswordToggle(!showPasswordToggle)}
      />
    ) : (
      <HideIconWrapper
        onClick={() => setShowPasswordToggle(!showPasswordToggle)}
      />
    );

  return (
    <SignUpContainer>
      <TitleContainer>I do not have an account</TitleContainer>
      <span>Sign up with your email and password</span>
      <form className="sign-up-form" onSubmit={handleSubmit}>
        <FormInput
          data-testid="displayName"
          type="text"
          name="displayName"
          value={newUser.displayName}
          handleChange={handleChange}
          label="Display Name"
          required
        />
        <Wrapper>
          <FormInput
            data-testid="email"
            type="email"
            name="email"
            value={newUser.email}
            handleChange={handleChange}
            label="Email"
            required
          />
          <ErrorMsgOnEmailAlreadyInUse emailAlreadyInUse={emailAlreadyInUse} />
        </Wrapper>
        <Wrapper>
          <FormInput
            data-testid="password"
            type={showPasswordToggle ? "text" : "password"}
            name="password"
            pattern={passwordRegex}
            value={newUser.password}
            handleChange={handleChange}
            label="Password"
            required
          />
          <ShowAndHidePasswordToggle />
          <PasswordStrenghtField password={newUser.password} />
        </Wrapper>
        <Wrapper>
          <FormInput
            type={showPasswordToggle ? "text" : "password"}
            data-testid="confirmPassword"
            name="confirmPassword"
            value={newUser.confirmPassword}
            handleChange={handleChange}
            label="Confirm Password"
            required
          />
          <ShowAndHidePasswordToggle />
          <ErrorMsgOnPasswordDoesntMatch
            passwordDontMatch={passwordDontMatch}
          />
        </Wrapper>
        <CustomButton
          data-testid="CustomButtonForSubmit"
          type="submit"
          style={{ width: "100%" }}
        >
          Sign up
        </CustomButton>
      </form>
    </SignUpContainer>
  );
}
const mapStateToProps = createStructuredSelector({
  userError: selectUserError,
});
const mapDispatchToProps = (dispatch) => ({
  signUpStart: (userCredentials) => dispatch(signUpStart(userCredentials)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SignUp);
