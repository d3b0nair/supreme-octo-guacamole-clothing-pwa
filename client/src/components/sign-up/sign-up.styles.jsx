import styled, { css } from "styled-components";
import { ReactComponent as VisibleIcon } from "../../assets/visible.svg";
import { ReactComponent as HideIcon } from "../../assets/hidden.svg";
export const SignUpContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 320px;
  @media screen and (max-width: 800px) {
    width: 100%;
    margin: 0 0 40px 0;
  }
`;

const iconCSS = css`
  position: absolute;
  right: 0.5em;
  top: 0em;
  cursor: pointer;
  fill: rgba(63, 98, 142, 0.5);
  &:hover {
    fill: #ff6e7f;
    transition: fill 0.6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
  }
`;
export const VisibleIconWrapper = styled(VisibleIcon)`
  ${iconCSS}
`;
export const HideIconWrapper = styled(HideIcon)`
  ${iconCSS}
`;
export const Wrapper = styled.div`
  position: relative;
`;
export const TitleContainer = styled.h2`
  margin: 10px 0;
`;
export const TextContainer = styled.div`
  background-color: #ff6e7f;
  width: 100%;
  padding: 5px 0;
  margin-top: -13%;
  margin-bottom: 15px;
  @media screen and (max-width: 320px) {
    margin-top: -14%;
  }
`;
const textSize = 16;
export const TextMsg = styled.span`
  display: block;
  font-size: ${textSize}px;
  @keyframes fadeIn {
    0% {
      display: none;
      opacity: 0;
    }

    1% {
      display: inline-block;
      opacity: 0;
    }

    100% {
      display: inline-block;
      opacity: 1;
    }
  }
  animation: fadeIn 1s ease-out;
`;
