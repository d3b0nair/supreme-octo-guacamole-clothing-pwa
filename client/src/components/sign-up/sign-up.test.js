import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import { SignUp } from "./sign-up.component";

describe("SignUp component", () => {
  const mockSignUpStart = jest.fn();
  const mockProps = (otherProps) => ({
    ...otherProps,
    signUpStart: mockSignUpStart,
  });
  const component = (mockProps) => <SignUp {...mockProps} />;

  afterEach(cleanup);

  it("should mount", () => {
    const { container } = render(component(mockProps()));
    expect(container).toMatchSnapshot();
  });
  it("should call mockSignUpStart when button clicked", async () => {
    render(component(mockProps()));
    await waitFor(() =>
      fireEvent.change(screen.getByTestId("displayName"), {
        target: { value: "username" },
      })
    );
    await waitFor(() =>
      fireEvent.change(screen.getByTestId("email"), {
        target: { value: "email@test.com" },
      })
    );
    await waitFor(() =>
      fireEvent.change(screen.getByTestId("password"), {
        target: { value: "password" },
      })
    );
    await waitFor(() =>
      fireEvent.change(screen.getByTestId("confirmPassword"), {
        target: { value: "password" },
      })
    );
    await waitFor(() =>
      fireEvent.click(screen.getByTestId("CustomButtonForSubmit"))
    );
    expect(mockSignUpStart).toHaveBeenCalled();
  });
});
