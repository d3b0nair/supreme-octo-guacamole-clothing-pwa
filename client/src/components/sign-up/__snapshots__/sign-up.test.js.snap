// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp component should mount 1`] = `
<div>
  <div
    class="sc-eCImPb zpltk"
  >
    <h2
      class="sc-furwcr bvQyCm"
    >
      I do not have an account
    </h2>
    <span>
      Sign up with your email and password
    </span>
    <form
      class="sign-up-form"
    >
      <div
        class="sc-bdvvtL hMBGla"
      >
        <input
          class="sc-gsDKAQ hVBime"
          data-testid="displayName"
          name="displayName"
          required=""
          type="text"
        />
        <label
          class="sc-dkPtRN jpMQjL  form-input-label"
          data-testid="FormInputLabel"
        >
          Display Name
        </label>
      </div>
      <div
        class="sc-iCfMLu FBMMr"
      >
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            class="sc-gsDKAQ hVBime"
            data-testid="email"
            name="email"
            required=""
            type="email"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Email
          </label>
        </div>
      </div>
      <div
        class="sc-iCfMLu FBMMr"
      >
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            class="sc-gsDKAQ hVBime"
            data-testid="password"
            name="password"
            pattern="^(?=.*)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$"
            required=""
            type="password"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Password
          </label>
        </div>
        <svg
          class="sc-gKclnd eHYgMs"
        >
          hidden.svg
        </svg>
      </div>
      <div
        class="sc-iCfMLu FBMMr"
      >
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            class="sc-gsDKAQ hVBime"
            data-testid="confirmPassword"
            name="confirmPassword"
            required=""
            type="password"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Confirm Password
          </label>
        </div>
        <svg
          class="sc-gKclnd eHYgMs"
        >
          hidden.svg
        </svg>
      </div>
      <button
        class="sc-hKwDye hzDzeB"
        data-testid="CustomButtonForSubmit"
        style="width: 100%;"
        type="submit"
      >
        Sign up
      </button>
    </form>
  </div>
</div>
`;
